crates/libp2p-networking/src/network/stake_table_transport.rs

```diff
@@ -0,0 +1,749 @@
+        if peer_id != *required_peer_id {
+            return Err(anyhow::anyhow!("Peer ID mismatch"));
+        }
+
+        // Check if the public key is in the stake table
+        if !stake_table.contains(&public_key) {
+            return Err(anyhow::anyhow!("Peer not in stake table"));
+        }
+    }
+
+    Ok(())
+}
+
+impl<T: Transport, S: SignatureKey + 'static, C: StreamMuxer + Unpin> Transport
+    for StakeTableAuthentication<T, S, C>
+where
+    T::Dial: Future<Output = Result<T::Output, T::Error>> + Send + 'static,
+    T::ListenerUpgrade: Send + 'static,
+    T::Output: AsOutput<C> + Send,
+    T::Error: From<<C as StreamMuxer>::Error> + From<std::io::Error>,
+
+    C::Substream: Unpin + Send,
+{
+    // `Dial` is for connecting out, `ListenerUpgrade` is for accepting incoming connections
+    type Dial = Pin<Box<dyn Future<Output = Result<T::Output, T::Error>> + Send>>;
+    type ListenerUpgrade = Pin<Box<dyn Future<Output = Result<T::Output, T::Error>> + Send>>;
+
+    // These are just passed through
+    type Output = T::Output;
+    type Error = T::Error;
+
+    /// Dial a remote peer. This function is changed to perform an authentication handshake
+    /// on top.
+    fn dial(
+        &mut self,
+        addr: libp2p::Multiaddr,
+    ) -> Result<Self::Dial, libp2p::TransportError<Self::Error>> {
+        // Perform the inner dial
+        let res = self.inner.dial(addr);
+
+        // Clone the necessary fields
+        let auth_message = Arc::clone(&self.auth_message);
+        let stake_table = Arc::clone(&self.stake_table);
+
+        // If the dial was successful, perform the authentication handshake on top
+        match res {
+            Ok(dial) => Ok(Box::pin(async move {
+                // Perform the inner dial
+                let mut stream = dial.await?;
+
+                // Time out the authentication block
+                async_timeout(AUTH_HANDSHAKE_TIMEOUT, async {
+                    // Open a substream for the handshake
+                    let mut substream =
+                        poll_fn(|cx| stream.as_connection().poll_outbound_unpin(cx)).await?;
+
+                    // (outbound) Authenticate with the remote peer
+                    authenticate_with_remote_peer(&mut substream, auth_message)
+                        .await
+                        .map_err(|e| {
+                            warn!("Failed to authenticate with remote peer: {:?}", e);
+                            std::io::Error::new(std::io::ErrorKind::Other, e)
+                        })?;
+
+                    // (inbound) Verify the remote peer's authentication
+                    verify_peer_authentication(&mut substream, stake_table, stream.as_peer_id())
+                        .await
+                        .map_err(|e| {
+                            warn!("Failed to verify remote peer: {:?}", e);
+                            std::io::Error::new(std::io::ErrorKind::Other, e)
+                        })?;
```

Comment:
I'm not sure how the `substream` here works but we effectively just always do things in this order (write our auth message, then wait for the peers and verify it) or does it some how mess up?  Just thinking that this logic is the same everywhere and would be nice if it were just one function called like `auth_with_peer(stream

Reply:
This one is actually the only one in this order, the others perform the reverse. We want to synchronously wait and make the handshake order explicit because if the peer is not actively receiving our message it will fail (e.g. if they are both sending data)

Reply:
Ok that's what I was afraid of.  I knew this one was the opposite order of the listening side so I wanted to be sure it had to be that order.  My thought was that sending could always happen first since both nodes need to get the auth message.  I was thinking they could both send at the same time, then both receive after sending but I guess one send might fail in that case